docker/ubuntu_16.04-i686.dockerfile

```diff
@@ -0,0 +1,28 @@
+FROM ubuntu:16.04
+
+ENV DEBIAN_FRONTEND noninteractive
+ENV PATH /root/.cargo/bin:$PATH
```

Comment:
As a general rule of thumb it's [strongly recommended to **not** run stuff inside the container as root](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#user). It's better to create a dedicated user and use it to execute things.

Reply:
I couldn't figure out a way to use that in combination with bind-mounts: if I switch to a non-root user, I lose access to the bind-mounted source.

My first idea was to use identical UIDs in both container and host, but that's obviously non-portable, and would only work on my machine. If we ever start distributing an actual Docker image (via Docker Hub, for example), it won't be universally useable.

My next idea was for container to copy the source someplace else, change permissions, and work on that. Apparently the only way to do this is to write an appropriate script, and ask users to run the script before any command (`docker run our-script.sh \&\& user-command`). Ugly and error-prone.

Do you see any other options, @der-lyse?

Reply:
After sleeping on it, I'm ready to take a step back and re-consider.

This Dockerfile is probably good enough for Cirrus CI. They're running a public cloud, so they should already be conscious of the risks they're subjecting themselves to. Since we're running only well-known FLOSS tools plus our own test suite, I don't think we increase that risk perceptibly.

With that said, this Dockerfile should be modified to *not* use root on CI if we find a way to do it. I can open an issue about that.

It can be argued that by the same token, this is safe to use locally, but luckily we can do better:
1. add `RUN useradd builder`, but don't add a `USER` directive—basically, reserve a user inside a container;
2. install `git` inside the container;
3. in our doc/internal/docker.md, encourage developers to `docker run --user=builder`, clone the code into the home directory, and build it—all as a unprivileged user.

Developers will have to push their changes somewhere before starting the container, but that's still a better way to test changes than what we have today—i.e. it's better than nothing :)